### PR TITLE
fix(idd-doctor): harden workshop back-link detection against false positives

### LINKED ISSUE #912: fix(idd-doctor): harden workshop back-link detection against false positives
Part of roadmap #910 (resolve unresolved merged-PR review feedback).

## Background

On PR #742, two review comments (Codex P2 + Copilot) on the
`containsExampleRepoBackLink` workshop back-link check were left
**unresolved** at merge. Both are still present on current `main`
(`src/scripts/idd-doctor.mts`; `scripts/idd-doctor.mjs` is generated) and
each produces a **false pass** of the `--require-github` back-link gate:

1. **List-marker lines inside an open indented code block are not treated
   as code.** In `stripIndentedCodeBlocksPreservingLines`, the condition
   `if (indented && (prevBlank || inBlock) && !looksLikeListItem)` applies
   the list-item exemption even in the `inBlock` continuation case. Per
   CommonMark §4.4 a list cannot start inside an open indented code block
   without an intervening blank line + dedent, so the line stays code.
   Effect: `paragraph\n\n    code line\n    - [link](workshop-url)` makes
   `containsExampleRepoBackLink(...)` return `true` even though no rendered
   back-link exists.

2. **Reference-style images are scanned as navigation links.**
   `containsExampleRepoBackLink` masks only inline images (`alt`);
   reference-style images (`![alt][id]` + `[id]: https://…`) are still
   consumed by the absolute-URL scan and the root-relative `mdRefDef`
   scan. Effect: a workshop URL used only as an image source
   (`![badge][b]\n\n[b]: https://github.com/kurone-kito/idd-skill/blob/main/docs/workshop/README.md`)
   returns `true`.

## Proposed change

In `src/scripts/idd-doctor.mts`:

- **(1)** Split the condition in `stripIndentedCodeBlocksPreservingLines`
  so the list-item exemption only protects a block *opener*: when
  `indented && inBlock`, always blank the line (continuation stays code);
  keep the `indented && prevBlank && !looksLikeListItem` branch to open a
  block.
- **(2)** In `containsExampleRepoBackLink`, collect labels referenced by
  images via `/!\[([^\]]*)\](?:\[([^\]]*)\])?/g` (covers full `![alt][id]`,
  collapsed `![id][]`, and shortcut `![id]`) into a lowercased `Set`, then
  exclude reference-definition lines whose label is in that set from both
  the absolute-URL scan and the `mdRefDef` scan. Keep real reference
  *links* (same label used by `[text][id]`, not an image) counting.
- Regenerate `scripts/idd-doctor.mjs` via `pnpm run build`.

## Acceptance criteria

- Add tests to `tests/idd-doctor.test.mts`:
  - code-block → indented list-marker line with a workshop URL → `false`
    (false-pass guard), plus a regression test that a real blank-separated
    nested list still → `true`.
  - reference-style image whose destination is a workshop URL → `false`
    (both absolute and root-relative forms), plus a real reference *link*
    → `true` (regression guard).
- No regression in the existing `containsExampleRepoBackLink` tests.
- `pnpm run lint` and `pnpm run build:check` pass.

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable; below the configured floor is human-oriented._

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

Two unresolved review findings (PR #742: Codex P2 + Copilot) on the
`--require-github` workshop back-link gate each produced a **false pass**
on current `main`. Both are fixed in `src/scripts/idd-doctor.mts`.

### 1. List-marker line inside an open indented code block

`stripIndentedCodeBlocksPreservingLines` applied the list-item exemption
even on the `inBlock` continuation case, so a line like
`    - [link](workshop-url)` continuing an open indented code block was
kept as a list item. Per CommonMark §4.4 a list cannot start inside an
open indented code block without an intervening blank line, so the
continuation now stays code and is blanked. The blank-separated-list and
not-blank-separated nested-list cases still resolve to a real list.

### 2. Reference-style image destinations scanned as links

`containsExampleRepoBackLink` masked only inline images, so a
reference-style image (`![alt][id]` + `[id]: <url>`) had its source
definition scanned by both the absolute-URL and root-relative ref-def
scans. Image-referenced labels (full `![alt][id]`, collapsed `![id][]`,
shortcut `![id]`) are now collected and their reference-definition lines
dropped before the URL scans. A real reference *link* (`[text][id]`)
keeps its definition and still counts.

## Changes

- `src/scripts/idd-doctor.mts`: both fixes above.
- `tests/idd-doctor.test.mts`: false-pass guards (code-block list-marker,
  reference-style image absolute + root-relative + shortcut) and
  regression guards (real reference link still counts).
- `scripts/idd-doctor.mjs`: regenerated via `pnpm run build`.

## Verification

`pnpm run lint` (966 tests, biome, dprint, cspell, markdownlint,
`audit-docs --check`) and `pnpm run build:check` pass.

Closes #912

Part of roadmap #910.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Markdown code block parsing to correctly preserve blank lines within indented code blocks.
  * Enhanced back-link detection to distinguish image references from actual navigational links in documentation.

* **Tests**
  * Added comprehensive test coverage for Markdown parsing edge cases and reference-style link/image handling scenarios.

### CODE REVIEW

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-82e2247e 2712d5d37a3d258d6269e38ed97063864b982767 2026-06-18T00:35:46Z

_claude-82e2247e: Copilot advisory-wait recovery marker — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai did not review HEAD 2712d5d (PR review rate limit reached, no review started); this is not a completed review and carries no advisory result to action.

File: src/scripts/idd-doctor.mts [hunk 1]
```diff
@@ -1549,7 +1549,18 @@ function stripIndentedCodeBlocksPreservingLines(content: string): string {
       inBlock = false;
       continue;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-doctor.mts:1560]: **Accepted** — fixed in 57c60af: keep `inBlock` across blank lines so an indented code block survives an internal blank line (CommonMark §4.4); a post-blank indented list-marker line now stays code. Added a regression test.

File: src/scripts/idd-doctor.mts [hunk 2]
```diff
@@ -1582,7 +1593,31 @@ export function containsExampleRepoBackLink(
...
+    if (label.length > 0) imageLabels.add(label.toLowerCase());
+  }
+  const stripped =
+    imageLabels.size === 0
+      ? imagedStripped
+      : imagedStripped.replace(
+          /^ {0,3}\[([^\]]+)\]:.*$/gm,
+          (full, label: string) =>
+            imageLabels.has(label.trim().toLowerCase()) ? '' : full,
+        );
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-doctor.mts:1693]: **Accepted** — fixed in 57c60af: a reference definition is now dropped only when its label is referenced exclusively by images; a label shared with a real reference link (`[text][id]`) keeps its definition so the navigation link still counts. Added a shared-label regression test.

**Comment by kurone-kito** [comment]
advisory-wait: claude-82e2247e 57c60af2adb65b16ac3f4b7cb11b16b53a22f8ad 2026-06-18T00:48:00Z

_claude-82e2247e: Copilot advisory-wait request marker — IDD automation marker. Do not edit._

File: src/scripts/idd-doctor.mts [hunk 3]
```diff
@@ -1544,12 +1544,28 @@ function stripIndentedCodeBlocksPreservingLines(content: string): string {
...
+      // here stays code and must be blanked too (it is not a list opener).
+      out.push('');
+      prevBlank = false;
+      continue;
+    }
+    if (indented && prevBlank && !looksLikeListItem) {
+      // Opener: an indented, non-list line after a blank line starts a
+      // new indented code block.
       out.push('');
       inBlock = true;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-doctor.mts]: **Accepted** — fixed in 0b54c68: `stripIndentedCodeBlocksPreservingLines` now tracks an open-list context. With no list open, a >=4-space indented list-marker line after a blank is treated as an indented code block (not a list), so its URL no longer false-passes; nested-list items under an open list are still preserved. Added a regression test.

File: src/scripts/idd-doctor.mts [hunk 4]
```diff
@@ -1582,7 +1598,53 @@ export function containsExampleRepoBackLink(
...
+  while ((linkMatch = refLinkPattern.exec(imagedStripped)) !== null) {
+    const explicit = linkMatch[1]?.trim() ?? '';
+    // Collapsed form `[id][]` carries the label in the first bracket;
+    // recover it from the full match when the second bracket is empty.
+    const label =
+      explicit.length > 0
+        ? explicit
+        : (linkMatch[0].match(/^\[([^\]]*)\]\[\]$/)?.[1]?.trim() ?? '');
+    if (label.length > 0) linkLabels.add(label.toLowerCase());
+  }
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-doctor.mts:1671]: **Accepted** — fixed in 0b54c68: link-label collection now also covers shortcut reference links (`[id]`), scoped to defined labels to avoid treating ordinary bracketed prose as a reference. A definition shared by an image and a shortcut link is therefore kept.

File: tests/idd-doctor.test.mts [hunk 5]
```diff
@@ -1014,6 +1014,64 @@ test('containsExampleRepoBackLink preserves links inside blank-separated nested
...
+  );
+});
+
+test('containsExampleRepoBackLink keeps counting a real reference-style link', () => {
+  // A reference *link* (no leading `!`) is navigation, so its definition
+  // is still scanned even though a reference *image* would be excluded.
+  const md =
+    'See the [workshop guide][w].\n\n[w]: https://github.com/kurone-kito/idd-skill/blob/main/docs/workshop/README.md';
+  assert.equal(containsExampleRepoBackLink(md, 'kurone-kito/idd-skill'), true);
+});
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply tests/idd-doctor.test.mts:1055]: **Accepted** — fixed in 0b54c68: added a regression test for a definition shared by a reference image and a shortcut reference link (expects the back-link to still count).

**Comment by kurone-kito** [comment]
advisory-wait: claude-82e2247e 0b54c6897451ace4456912c49190d9f142781a3e 2026-06-18T01:00:44Z

_claude-82e2247e: Copilot advisory-wait request marker — IDD automation marker. Do not edit._

File: src/scripts/idd-doctor.mts [hunk 6]
```diff
@@ -1544,19 +1550,52 @@ function stripIndentedCodeBlocksPreservingLines(content: string): string {
...
+        out.push('');
+        prevBlank = false;
+        continue;
+      }
+      if (listOpen) {
+        // Indented content under an open list is list continuation or a
+        // nested list item, not a top-level code block: keep it.
+        out.push(line);
+        prevBlank = false;
+        continue;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/idd-doctor.mts:1576]: **Rejected** — deferred to #945 (reasoned rejection, not a denial of validity). This is a genuine but contrived residual: a workshop URL inside an indented code block *nested under a list item*. The proposed `>=8 spaces under a list = code` threshold trades one approximation for another (it mis-classifies deeply nested list content and ordered-list bodies), and distinguishing nested code blocks from list continuation correctly requires tracking each list item content-indent column — effectively a CommonMark block parser, which is out of scope for #912 (whose two named bugs and the review-driven open-list/shortcut refinements are now fixed). Tracked as a standalone follow-up in #945 with the regression test you requested.

**Comment by kurone-kito** [comment]
_claude-82e2247e: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-82e2247e: critique baseline — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai posted only a rate-limit notice (re-edited in place at 00:59:47Z) and did not complete a review of HEAD 0b54c68; this is not a completed review and carries no advisory result to action. Re-affirming the disposition after the in-place edit. Copilot completed its review of this HEAD and CI is green.

**Comment by kurone-kito** [comment]
_claude-82e2247e: review triage snapshot — IDD automation marker. Do not edit._
